src/Service.Tests/Unittests/MultipleCreateUnitTests/MultipleCreateOrderHelperUnitTests.cs

```diff
@@ -0,0 +1,399 @@
+        /// all the columns required to do insertion and act as a referenced entity.
+        ///
+        /// Here, the relationship between entities: 'User_AutogenToNonAutogenRelationshipColumn' and
+        /// 'UserProfile_NonAutogenToAutogenRelationshipColumn' is defined as:
+        /// User_AutogenToNonAutogenRelationshipColumn(userid, username) -> UserProfile_NonAutogenToAutogenRelationshipColumn(userid, username)
+        /// where both the relationships fields User_AutogenToNonAutogenRelationshipColumn.userid is an autogenerated field while all other
+        /// relationship fields are non-autogenerated.
+        /// </summary>
+        [TestMethod]
+        public void ValidateDeterministicReferencingEntityForAutogenAndNonAutogenRelationshipColumns()
+        {
+            // Test 1
+            IMiddlewareContext context = SetupMiddlewareContext();
+            string sourceEntityName = "User_AutogenToNonAutogenRelationshipColumn";
+            string targetEntityName = "UserProfile";
+
+            // Setup column input in source entity.
+            Dictionary<string, IValueNode> columnDataInSourceBody = new()
+            {
+                { "email", new StringValueNode("dab@microsoft.com") },
+                { "username", new StringValueNode("DAB") }
+            };
+
+            // Setup column input in target entity.
+            ObjectValueNode targetNodeValue = new();
+            List<ObjectFieldNode> fields = new()
+            {
+                new ObjectFieldNode("profilepictureurl", "dab/profilepicture")
+            };
+
+            targetNodeValue = targetNodeValue.WithFields(fields);
+
+            string referencingEntityName = MultipleCreateOrderHelper.GetReferencingEntityName(
+                relationshipName: "UserProfile_AutogenToNonAutogenRelationshipColumn",
+                context: context,
+                sourceEntityName: sourceEntityName,
+                targetEntityName: targetEntityName,
+                metadataProvider: _sqlMetadataProvider,
+                columnDataInSourceBody: columnDataInSourceBody,
+                targetNodeValue: targetNodeValue,
+                nestingLevel: 1);
+
+            Assert.AreEqual(targetEntityName, referencingEntityName);
+        }
+
+        private static IMiddlewareContext SetupMiddlewareContext()
+        {
+            Mock<IMiddlewareContext> context = new();
+            Mock<IVariableValueCollection> variables = new();
+            context.Setup(x => x.Variables).Returns(variables.Object);
+            return context.Object;
+        }
+
+        /// <summary>
+        /// Test to validate that when an FK constraint exists between the entities, we can determine which of the entity
+        /// acts as the referencing entity based on the database metadata that we collect at the startup.
+        /// The entity which holds the foreign key reference acts as the referencing entity.
+        /// </summary>
```

Comment:
be explicit that this test method exercises: `MultipleCreateOrderHelper.GetReferencingEntityName(...)` because that function is abstracted away in a `Validate..(..)` helper function.